client/consul/consul.go

```diff
@@ -29,3 +36,101 @@ type ServiceIdentityAPI interface {
 type SupportedProxiesAPI interface {
 	Proxies() (map[string][]string, error)
 }
+
+// JWTLoginRequest is an object representing a login request with JWT
+type JWTLoginRequest struct {
+	JWT            string
+	Role           string
+	AuthMethodName string
+}
+
+// ConsulClient is the interface that the nomad client uses to interact with
+// Consul.
+type ConsulClient interface {
+	// DeriveSITokenWithJWT contacts the nomad server and fetches wrapped tokens
+	// for a set of tasks. The wrapped tokens will be unwrapped using consul and
+	// returned.
+	DeriveSITokenWithJWT(map[string]JWTLoginRequest) (map[string]string, error)
+}
+
+type consulClient struct {
+	// tokenDeriver is a function pointer passed in by the client to derive
+	// tokens by making RPC calls to the nomad server.
+	tokenDeriver TokenDeriverFunc
+
+	// client is the API client to interact with vault
+	client *consulapi.Client
+
+	// config is the configuration to connect to consul
+	config *config.ConsulConfig
+
+	logger hclog.Logger
+}
+
+// NewConsulClient creates a new Consul client
+func NewConsulClient(config *config.ConsulConfig, logger hclog.Logger, tokenDeriver TokenDeriverFunc) (*consulClient, error) {
+	if config == nil {
+		return nil, fmt.Errorf("nil consul config")
+	}
+
+	logger = logger.Named("consul")
+
+	c := &consulClient{
+		config:       config,
+		logger:       logger,
+		tokenDeriver: tokenDeriver,
+	}
+
+	// if UseIdentity is unset of set to false, return an empty client
+	if config.UseIdentity == nil || !*config.UseIdentity {
+		return nil, nil
+	}
+
+	// Get the Consul API configuration
+	apiConf, err := config.ApiConfig()
+	if err != nil {
+		logger.Error("error creating default Consul API config", "error", err)
+		return nil, err
+	}
+
+	// Create the API client
+	client, err := consulapi.NewClient(apiConf)
+	if err != nil {
+		logger.Error("error creating Consul client", "error", err)
+		return nil, err
+	}
+
+	useragent.SetHeaders(client)
+	c.client = client
+
+	return c, nil
+}
+
+// DeriveSITokenWithJWT takes a JWT from request and returns a consul token for
+// each workload in the request
+func (c *consulClient) DeriveSITokenWithJWT(reqs map[string]JWTLoginRequest) (map[string]string, error) {
+	tokens := make(map[string]string, len(reqs))
+	var mErr *multierror.Error
+
+	for k, req := range reqs {
+		// login using the JWT and obtain a Consul ACL token for each workload
+		t, _, err := c.client.ACL().Login(&consulapi.ACLLoginParams{
+			AuthMethod:  req.AuthMethodName,
+			BearerToken: req.JWT,
+		}, &consulapi.WriteOptions{})
+		if err != nil {
+			mErr = multierror.Append(mErr, fmt.Errorf(
+				"failed to authenticate with consul for identity %s: %v", k, err,
+			))
+			continue
+		}
+
+		tokens[k] = t.SecretID
+	}
+
+	if err := mErr.ErrorOrNil(); err != nil {
+		return nil, err
```

Comment:
Is there any value in returning `tokens, err` so that we can return _some_ tokens if we only get an error on one of the logins?

Reply:
good point! Fixed in [4c925f3](https://github.com/hashicorp/nomad/pull/18370/commits/4c925f3a61b8b612482abe61fc7013dab06fd041)